### PR TITLE
Tests for podman image scp (the sudo form)

### PR DESCRIPTION
**Pull request by edsantiago**
Start inching our way back to having tests for the sudo form
of podman image scp. Basically, copy an image to another user
and then back, using a pseudorandom name. Confirm that the
image makes it to the remote end, and that when we copy it
back, the original image digest is preserved.

When scp'ing as root, we identify the destination rootless
user account via the $PODMAN_ROOTLESS_USER envariable. Setting
this and creating the account is left as an exercise for the
CI framework (be it github, or Fedora/CentOS/RHEL gating, or
other).

Also: amend hack/bats to set and relay $PODMAN_ROOTLESS_USER,
so developers can test locally.

Also: remove what I'm 99% sure is a debugging printf.

Signed-off-by: Ed Santiago <santiago@redhat.com>

### CODE REVIEW

**Comment by edsantiago** [comment]
Root-to-rootless is _currently untested_. It's much more complicated than I had expected, because our Cirrus CI only creates the rootless account on-demand, just before using it to run rootless tests. We need to amend  https://github.com/containers/podman/blob/92cd152b533fc39fa698644e0fa41716e6999e52/contrib/cirrus/setup_environment.sh#L185-L195 so it _always_ runs `setup_rootless()`, every time, and sets `$PODMAN_ROOTLESS_USER` (note the `PODMAN_` prefix). This is left as an exercise for tomorrow.

Root-to-rootless is also not working anyway, at least not on my laptop:
```
   # .../podman tag quay.io/libpod/testimage:20210610 foo.bar/nonesuch/c_tckzsgu9nu:mytag
   # .../podman image scp foo.bar/nonesuch/c_tckzsgu9nu:mytag esm@localhost::
   Copying blob f36118df491f done
   Copying config 9f9ec7f2fd done
   Writing manifest to image destination
   Storing signatures
   Error: payload does not match any of the supported image formats:
    * oci: initializing destination containers-storage:[overlay@/home/esm/.local/share/containers/storage+/run/user/14904/containers]localhost/var/tmp/podman541448797:latest: creating a temporary directory: stat /this/is/set/in/containers.conf: no such file or directory
    * oci-archive: loading index: open /var/tmp/oci445817820/index.json: no such file or directory
    * docker-archive: initializing destination containers-storage:[overlay@/home/esm/.local/share/containers/storage+/run/user/14904/containers]foo.bar/nonesuch/c_tckzsgu9nu:mytag: creating a temporary directory: stat /this/is/set/in/containers.conf: no such file or directory
    * dir: open /var/tmp/podman541448797/manifest.json: not a directory
```
Again, I'm done worrying for today. The problem will still be there tomorrow.

File: test/system/120-load.bats
Comment (edsantiago) [inline test/system/120-load.bats:128]: Leaving this in your capable hands, @cdoern

**Comment by edsantiago** [comment]
Rootless [f34](https://storage.googleapis.com/cirrus-ci-6707778565701632-fcae48/artifacts/containers/podman/4993076753072128/html/sys-podman-fedora-34-rootless-host.log.html#t--00166), [f35](https://storage.googleapis.com/cirrus-ci-6707778565701632-fcae48/artifacts/containers/podman/6400451636625408/html/sys-podman-fedora-35-rootless-host.log.html#t--00166), and [ubuntu](https://storage.googleapis.com/cirrus-ci-6707778565701632-fcae48/artifacts/containers/podman/6118976659914752/html/sys-podman-ubuntu-2110-rootless-host.log.html#t--00166) all ran and passed. (As opposed to getting skipped). That was the purpose of this WIP run: making sure that all rootless environments are able to `sudo` to root.

I haven't looked at root tests, because I know those were all skipped. Remote, too. This CI run has served its purpose. Tomorrow, we look at sudo'ing from root to rootless.

**Comment by cdoern** [comment]
@edsantiago please apply this patch for the quiet issue, it will still print the load result as `podman load` does: [scp.patch.txt](https://github.com/containers/podman/files/7843094/scp.patch.txt)

File: cmd/podman/images/scp.go
Comment (cdoern) [inline cmd/podman/images/scp.go:214]: this was meant as a permanent output to tell the user some concrete information about what they transferred and who they transferred to, @rhatdan would you be opposed to removing this? I think having some sort of scp specific output is a good idea.

File: test/system/120-load.bats
Comment (cdoern) [reply test/system/120-load.bats:128]: patch added below

**Comment by cdoern** [comment]
> Root-to-rootless is _currently untested_. It's much more complicated than I had expected, because our Cirrus CI only creates the rootless account on-demand, just before using it to run rootless tests. We need to amend
> 
> https://github.com/containers/podman/blob/92cd152b533fc39fa698644e0fa41716e6999e52/contrib/cirrus/setup_environment.sh#L185-L195
> so it _always_ runs `setup_rootless()`, every time, and sets `$PODMAN_ROOTLESS_USER` (note the `PODMAN_` prefix). This is left as an exercise for tomorrow.
> 
> Root-to-rootless is also not working anyway, at least not on my laptop:
> 
> ```
>    # .../podman tag quay.io/libpod/testimage:20210610 foo.bar/nonesuch/c_tckzsgu9nu:mytag
>    # .../podman image scp foo.bar/nonesuch/c_tckzsgu9nu:mytag esm@localhost::
>    Copying blob f36118df491f done
>    Copying config 9f9ec7f2fd done
>    Writing manifest to image destination
>    Storing signatures
>    Error: payload does not match any of the supported image formats:
>     * oci: initializing destination containers-storage:[overlay@/home/esm/.local/share/containers/storage+/run/user/14904/containers]localhost/var/tmp/podman541448797:latest: creating a temporary directory: stat /this/is/set/in/containers.conf: no such file or directory
>     * oci-archive: loading index: open /var/tmp/oci445817820/index.json: no such file or directory
>     * docker-archive: initializing destination containers-storage:[overlay@/home/esm/.local/share/containers/storage+/run/user/14904/containers]foo.bar/nonesuch/c_tckzsgu9nu:mytag: creating a temporary directory: stat /this/is/set/in/containers.conf: no such file or directory
>     * dir: open /var/tmp/podman541448797/manifest.json: not a directory
> ```
> 
> Again, I'm done worrying for today. The problem will still be there tomorrow.

I think yours fails due to limitations of podman save only accepting certain formats. I will look into this tomorrow as well to see if there is a way I can reformat the image on the fly.

**Comment by edsantiago** [comment]
Thanks for the `-q` patch. With it, output is just:
```console
$ bin/podman image scp -q quay.io/libpod/testimage:20210610 root@localhost::
Loaded image(s): quay.io/libpod/testimage:20210610
```
...which I'll confess to being ambivalent about. I don't have a good sense for what `-q` output should look like, but I would expect something more like:
```
root@localhost::quay.io/libpod/testimage:20210610
```
...or perhaps even no output at all. I think this merits wider discussion.

As for the SOURCE/DEST, I don't see those providing any value whatsoever to anyone -- I genuinely thought they were just debugging printfs that you forgot to remove. For reference, here are command results as they stand today:
```console
$ bin/podman image scp -q quay.io/libpod/testimage:20210610 root@localhost::
Copying blob f36118df491f done
Copying config 9f9ec7f2fd done
Writing manifest to image destination
Storing signatures
Getting image source signatures
Copying blob f36118df491f skipped: already exists
Copying config 9f9ec7f2fd done
Writing manifest to image destination
Storing signatures
Loaded image(s): quay.io/libpod/testimage:20210610
SOURCE: {
    "file": "/dev/shm/podman120188115",
    "image": "quay.io/libpod/testimage:20210610",
    "user": "esm"
}
DEST: {
    "file": "/dev/shm/podman120188115",
    "user": "root"
}
```
I find that uncomfortably verbose. I accept your point about offering scp-specific output, but I really don't think that's it.

**Comment by rhatdan** [comment]
Yes I think this should become less verbose.

$ scp  dwalsh@localhost:/tmp/dan /tmp/walsh
dan                                                                                 100%    0     0.0KB/s   00:00    

So
```
$ podman image scp quay.io/libpod/testimage:20210610 root@localhost::
Copying blob f36118df491f done
Copying config 9f9ec7f2fd done
Writing manifest to image destination
Storing signatures
Getting image source signatures
Copying blob f36118df491f skipped: already exists
Copying config 9f9ec7f2fd done
Writing manifest to image destination
Storing signatures
quay.io/libpod/testimage:20210610    SIZE
$ podman image scp -q quay.io/libpod/testimage:20210610 root@localhost::
quay.io/libpod/testimage:20210610    SIZE
```

**Comment by edsantiago** [comment]
Today's non-progress:
* in enabling the rootless account, root's ssh authorized_hosts got set up; which then
* made the `podman system connection - ssh` test decide to run (it has been skipped since inception); which then
* failed all podman-remote tests, because `podman system connection` does not work.

I know. It works on Macs, maybe, or at least it's supposed to. It does not work on Linux. Not root nor rootless. I've spent most of today on an f35 VM trying every combination I can (identity file, agent, command-line permutations).
```console
# podman system service -t 0 unix:/tmp/foo &
[1] 22873
# ssh-agent bash
# ssh-add
Identity added: /root/.ssh/id_rsa (/root/.ssh/id_rsa)
# ssh localhost date
Tue Jan 11 05:27:44 PM EST 2022
# podman-remote system connection add --socket-path /tmp/foo mycon localhost
# podman-remote --log-level=debug info
INFO[0000] podman-remote filtering at log level debug
DEBU[0000] Called info.PersistentPreRunE(podman-remote --log-level=debug info)
DEBU[0000] Found SSH_AUTH_SOCK "/tmp/ssh-XXXXXXM82z3s/agent.22915", ssh-agent signer(s) enabled
DEBU[0000] SSH Agent Key SHA256:9j1blwt3wcrRiGYZQ7ZGu9axm3cDklH6/z4c+Ee8CzE ssh-rsa
Cannot connect to Podman. Please verify your connection to the Linux system using `podman system connection list`, or try `podman machine init` and `podman machine start` to manage a new Linux VM
Error: unable to connect to Podman. failed to create sshClient: Connection to bastion host (ssh://root@localhost:22/tmp/foo) failed.: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain
```
The socket is fine:
```console
# podman-remote --url unix:/tmp/foo info
...works fine
```
Tomorrow is another day, and this will wait.

File: test/system/120-load.bats
Comment (cdoern) [inline test/system/120-load.bats:104]: do you mean change the graph root here? can you not just do `run_podman --root=....` also I have seen some tests that use a different podman call rather than run_podman, could be a solution

Comment (edsantiago) [reply test/system/120-load.bats:104]: I meant, some way to have the `sudo`'ed podman have a custom `--root`, so that I, the userid running tests, do not cause harm to a different user. They can't both have the same root, of course; and there's no actual use to scp'ing to/from a different root on the test user. So it would be something like `scp myimage root@localhost[--root=/tmp/fakeroot]` which has no conceivable purpose other than testing. So I should really just delete that comment!

File: test/system/120-load.bats
Comment (edsantiago) [reply test/system/120-load.bats:132]: Thanks. I will pause work on this PR, and wait for a solution, because working around machinectl plus the warnings on Ubuntu is proving time-consuming.

**Comment by edsantiago** [comment]
@containers/podman-maintainers @cdoern @cevich PTAL

I've removed the WIP label and would like to merge this ASAP even though it's complete rubbish. Reason:  #12867 is in rapid churn, but it's being submitted without any testing whatsoever, and every iteration I've (manually) tested has been completely nonfunctional. I believe it is urgent that #12867 be blocked until this merges, and then rebase on this, and then fix all (ok, some) of the horrible FIXMEs in here but at least get _some_ testing.

**Comment by rhatdan** [comment]
Label still says [WIP]?

**Comment by edsantiago** [comment]
Oops. Fixed, thanks.

**Comment by rhatdan** [comment]
LGTM
Will leave thisto @cdoern to commit.

**Comment by edsantiago** [comment]
I'd like @cevich to review also, since this mucks with the Cirrus setup

File: test/system/120-load.bats
Comment (cdoern) [inline test/system/120-load.bats:83]: ubuntu gave me some issues on my first go as well, I can circle back in a separate PR and try to fix this

**Review by cdoern — APPROVED** [review]
LGTM, just a general comment about future ubuntu work. I don't have approve or merging ability I don't think.

File: test/system/120-load.bats
Comment (cdoern) [reply test/system/120-load.bats:83]: my PR no longer uses machinectl to execute the commands but does use it to log in which is an issue... I am looking into how runuser can be used instead in the rootful form (which is all that matters in this case since rootless just uses an exec.Command(podman...))

**Comment by cevich** [comment]
> I'd like @cevich to review also, since this mucks with the Cirrus setup

So...these changes are okay.  I guess my only (minor) issue is possible confusion due to "All-roads lead to `setup_rootless()`", so it could just be called from the top-level instead of every `case` statement.

**Comment by edsantiago** [comment]
Sorry about the confusion: all roads _do not_ lead to `setup_rootless`. In root, only when tests are `sys`. I tried calling it from the top level but that failed super early, it didn't even make it to tests. I don't remember the failure, but it was maybe in smoke or validate or one of the very early steps.

**Comment by cevich** [comment]
Ahh ok, well I'll defer to your actual experience then, rather than my hypothesized guesses :grinning:

**Pull request merged**
